src/pages/learn/mutations.mdx

```diff
@@ -0,0 +1,149 @@
+import { Callout } from "nextra/components"
+
+# Mutations
+
+<p className="learn-subtitle">Learn how to modify data with a GraphQL server</p>
+
+Most discussions of GraphQL focus on data fetching, but any complete data platform needs a way to modify server-side data as well.
+
+In REST, any request might cause some side-effects on the server, but by convention, it's suggested that one doesn't use `GET` requests to modify data. GraphQL is similar—technically any query could be implemented to cause a data write. However, it's useful to establish a convention that any operations that cause writes should be sent explicitly via a mutation.
+
+On this page, you'll learn how to use mutation operations to create, update, and delete data using GraphQL.
```

Comment:
I am very uncomfortable with this positioning; GraphQL isn't just a CRUD system and in general it's better to use semantically named mutations for specific user actions rather than trying to fit every mutation into the CRUD box. Some mutations such as `processPayment`, `resetPassword`, `sendEmailVerification`, `inviteUser`, `markAllNotificationsAsRead`, `redeemGiftCard` and so on (arguably) don't even fit into the CRUD box. (This isn't to say that CRUD should never be used, but more that a semantically named action should be exposed preferentially where it makes sense to do so.)

From GraphQL's point of view there are no different types of mutation - there is no difference between a create, update, delete or arbitrary user action mutation - so I'm concerned that highlighting CRUD mutations in this way sets the reader (especially someone who does not read the page fully) up for misconceptions.

Reply:
I see where you're coming from. I think the challenge here is to figure out how much information to include up-front in the intro for a reader who has some experience with REST and needs to connect those dots with GraphQL. This distinction is called out in the paragraph that follows the new example for updating a human name, but it's halfway down the page.

Do you think it would be better to generalize it like this?

> On this page, you'll learn how to use mutation operations to modify data on a server using GraphQL.

Reply:
Do you think the subheads on the page are problematic? We could further deemphasize any kind of 1:1 CRUD implication by renaming them like this:

~Create data~ Add new data
~Update data~ Update existing data
~Delete data~ Remove existing data

Reply:
One more thought on this is—I think there's a gap in the Best Practices content where "migrating from REST" could be covered in more detail too with a few more detailed examples that are based on a further-enhanced version of the Star Wars schema.

Reply:
I'm not sure that our learn docs should necessarily cater primarily to people who already know REST/CRUD - perhaps a separate guide for "how to convert CRUD REST operations to GraphQL" might make sense but I think our learn docs shouldn't start with REST concepts and map them to GraphQL, but instead start with GraphQL best practices and then if necessary expand into REST accommodations.

I think my preferred way of addressing this would be to start with a non-CRUD mutation that follows best practices (input object, mutation payload) so that that is front and centre, and then maybe have a sub heading about CRUD that has examples of create, update, upsert, delete. I think the original documentation used a create mutation because that's fairly universal, but perhaps pivoting away from that would help learners to do GraphQL the right way from the start?

Trying to think of something within the Star Wars schema that would make sense as a non-CRUD mutation, perhaps `rateFilm` which stores a rating "onto the film" from your perspective, but clearly isn't actually updating the film itself:

```graphql
type Mutation {
  rateFilm(input: RateFilmInput!): RateFilmPayload
}

input RateFilmInput {
  filmId: ID!
  rating: FilmRating! = THUMBS_UP
}

type RateFilmPayload {
  film: Film
}

type Film {
  # ...
  viewerRating: FilmRating
}

enum FilmRating {
  THUMBS_DOWN
  NEUTRAL
  THUMBS_UP
  DOUBLE_THUMBS_UP
}
```

Reply:
Actually I think this page should talk about the input object/payload best practices at the end, so maybe simplify this down:

```graphql
type Mutation {
  rateFilm(
    filmId: ID!
    rating: FilmRating! = THUMBS_UP
  ): Film
}

type Film {
  # ...
  viewerRating: FilmRating
}

enum FilmRating {
  THUMBS_DOWN
  NEUTRAL
  THUMBS_UP
  DOUBLE_THUMBS_UP
}
```

Reply:
I can come up with something but at this point I think I'll wait until #1812 is merged so I rebase the changes made in `swapi-schema.tsx` in that PR first.

Reply:
This is now merged :+1:

Reply:
There's a new commit with the requested changes included now.